### DISCUSSION TITLE
GPU ultilisation

### DISCUSSION THREAD

**Discussion by sannatitus** [bonsai-rx/bonsai#1540] [Q&A]
Hello! 
Does anyone have advice on how to ensure Bonsai is utilising my GPUs, rather than my desktop's RAM? 
I am using 2x identical Logitech Brio 4K webcams to record from two grouped nodes in the same Bonsai workflow. The two grouped nodes are identical for each camera, besides of course the port numbers. On the Logitech SDK (Logitech Capture v 02.08.11, downloaded from https://www.logitech.com/en-gb/software/capture.html ), settings are also identical among both webcams (4096x2160 resolution, writing at 30FPS with anti-flicker at 50FPS, exposure, etc.). Viewing on Logitech's SDK, the quality of both webcams looks identical. When running the Bonsai workflow, one camera has consistently lower resolution, flickering FPS and shorter videos (despite camera capture & video writer & Logitech SDK all acquiring at 30FPS for both cameras). 
This process slowly drives up my memory (128GB), to the point where I either cannot do anything else on the computer, the Bonsai workflow crashes or the entire computer crashes. I have tried to change settings in the Nvidia GeForce Experience, and everything looks ok? My IT department is looking into overclocking the GPU, but I'm not sure if this is sustainable for a 24/7 workflow which has no intention of stopping. Chatting with the lovely Jai Bhagat, he mentioned it's possible it's a Bonsai issue; potentially, my 'zip' node is oversubscribing to camera frames and driving up memory. 
Has anyone else experienced this and has a solution?
Attached is my workflow (I'm aware the 'tub' camera capture node is set to a lower acquiring resolution; if I set it to 4K to match the Logitech SDK, the workflow says 'it cannot acquire camera frame'). 
Thank you for your time,
Sanna 
[tank+tub_daily.zip](https://github.com/bonsai-rx/bonsai/files/12474683/tank%2Btub_daily.zip)

**Comment by bruno-f-cruz** [comment]
Heya :)

1) Not sure what you mean by using GPU. None of your operators requires the use of a GPU. What would you like to use your GPU for? Encoding the video?

2) Regarding the memory leak, i dont believe is related with the zip, as all your events going into the zip operator seem to be synchronous. The likely culprit is your VideoWriter node that it is not able to encode frames as fast as you acquire them, and it will just queue them in RAM, eventually blowing up at some point.

3) Where are you setting "(4096x2160 resolution, writing at 30FPS with anti-flicker at 50FPS, exposure, etc.). "? Since you are using a generic camera node, I am not sure I see how you can access all these settings from bonsai. You might also want to take a look at "VideoCaptureDevice" as it might expose several more features of the camera. (Although, this operator will drop frames if your workflow cannot keep up with the encoding of the video, just FYI, so that problem must still be solved)

Cheers,
Bruno

**Comment by sannatitus** [comment]
Hello Bruno the Bonsai wizard! What an honour. 

1. We have the CameraCapture output property Image node open when we are writing daily videos, so we can view our animals without disturbing them. I thought viewing 2x this node’s feed while writing the videos would involve the GPU (perhaps my limited understanding of this is incorrect), rather than relying on just CPU & memory. My IT department & I thought pushing more processes to rely on GPU would sort our memory leak! Good to know that’s maybe not the right direction. 
2. Any way to sort this within VideoWriter? A different codec perhaps? I’m very flexible on codec & container, just as long as it’s 4K. I’m even happy with 1080P if necessary, I just feel that there is no reason why this desktop shouldn’t be able to handle 2x 4K webcams. 
3. Those settings are in the Logitech SDK. In Bonsai’s CameraCapture however, you can expand the (+) properties to acquire at a certain resolution - that’s where I’ve set the 4K on one tank and 1080P on the other. The other settings I mentioned are in the Bonsai VideoWriter (FPS, resolution again, codec). 

I’d like to avoid dropping frames and queuing up this memory leak. What’s the best next step(s) to try? 

Cheers,
Sanna

**Reply by roaldarbol** [reply]
@sannatitus Did you find a solution to this? I am now trying to record videos(with substantially less CPU) and am running into this issue. I've also managed quite well previously in other software too (even on Raspberry Pi's), so this comes as a surprise to me.

**Reply by roaldarbol** [reply]
My current workaround is to save repeatedly, but even then, I have to save every hour (see #1597).

**Reply by roaldarbol** [reply]
Hi again! I tried monitoring the memory, and it seems that the memory leak is in fact in the streaming window (i.e., when you have watch the stream). When you close those windows, the memory leak disappears, and I successfully recorded 20h long videos from 2 cameras at 60fps on a 8GB RAM Windows desktop overnight. I'll report the leak as an issue later today when I have time. :-)

**Comment by sannatitus** [comment]
Hiya, no, unfortunately we haven't found a solution for this. The workstation was having some misc issues so it's being diagnosed at the manufacturers. We will attempt a new Bonsai workflow when it returns. Our memory leak doesn't stem from viewing the streaming window; the computer still crashed even with only the main bonsai window open. We might attempt using a HARP to trigger the cameras, so the computer doesn't get overwhelmed with too many frames(?). If we find a solution without requiring an external signal though (which should be possible!), I'll update it here.

**Reply by roaldarbol** [reply]
Ah damn. Hope you'll find a solution!
